src/combinatorics/ramsey/infinite.lean

```diff
@@ -0,0 +1,468 @@
+instance : has_mem ℕ Inf := ⟨λ n H, n ∈ H.s⟩
+instance : has_subset Inf := ⟨λ H₁ H₂, H₁.s ⊆ H₂.s⟩
+
+open color
+
+/--
+Given a function from natural numbers to colors,
+this property describes the numbers which are mapped to red.
+-/
+def reds (f : ℕ → color) (H : set ℕ) := {n : ℕ | f n = red ∧ n ∈ H}
+
+/--
+Given a function from natural numbers to colors,
+this property describes the numbers which are mapped to blues.
+-/
+def blues (f : ℕ → color) (H : set ℕ) := (λ n : ℕ, f n = blue ∧ H n)
+
+lemma lt_succ_sum (x y : ℕ) : x < x + y + 1 :=
+  lt_add_of_pos_right x (succ_pos y)
+
+/--
+Unordered pairs are defined as ordered pairs with the condition
+that the second element is strictly greater than the first.
+-/
+def unordered_pairs (S : set ℕ) :=
+  {p : ℕ × ℕ // p.fst < p.snd ∧ p.fst ∈ S ∧ p.snd ∈ S}
+
+local notation `[ℕ]²` := unordered_pairs univ
+local notation `[` S `]²` := unordered_pairs S
+
+/--
+Given a number m and a coloring f, the projection of a coloring
+of unordered_pairs to a coloring of numbers is given by fixing
+the first element as m. Note that red is arbitrarily chosen
+as the color assigned to any number below m.
+-/
+def project (f : [ℕ]² → color) (m : ℕ) : ℕ → color :=
+  λ n : ℕ,
+    if h : m < n
+      then f ⟨(m, n), ⟨h, trivial, trivial⟩⟩
+      else red
+
+/--
+The coloring of an unordered pair {a, b}, where a < b,
+agrees with the projection of a applied to b.
+-/
+@[simp] lemma project_eq (f : [ℕ]² → color) (p : [ℕ]²):
+  project f p.val.fst p.val.snd = f p :=
+begin
+  unfold project,
+  simp [p.property.left]
+end
+
+/--
+Given coloring of natural numbers, a homogeneous set is
+one which is always mapped to the same color.
+-/
+def homogeneous (f : ℕ → color) (H : set ℕ) :=
+  ∃ c : color, ∀ n, n ∈ H → f n = c
+
+/--
+Given a coloring f, a homogeneous projection is a set of natural numbers
+together with a single natural number.
+The intended use is a set which is homogeneous when projected
+with the given number and coloring.
+-/
+structure homogeneous_proj (f : [ℕ]² → color) extends Inf := (pt : ℕ)
+
+instance (f : [ℕ]² → color) : has_coe (homogeneous_proj f) Inf := ⟨λ c, c.to_Inf⟩
+
+/--
+A homogeneous projection p is refined by another homogeneous projection q if:
+the number given by q is greater than the number given by p,
+the number given by q is an elment of the set given by p,
+the set given by q is contained in the set given by p,
+and the projection using the number given by q is homogeneous
+on the set given by q.
+-/
+def refines {f : [ℕ]² → color} (p q: homogeneous_proj f) :=
+p.pt < q.pt ∧ q.pt ∈ p.s ∧ q.s ⊆ p.s ∧ homogeneous (project f q.pt) q
+
+instance (f : [ℕ]² → color) : has_lt (homogeneous_proj f) :=
+  ⟨λp q : homogeneous_proj f, refines p q⟩
+
+local attribute [instance] prop_decidable
+
+/--
+A version of the infinite pigeon hole principle, tailored to our use case.
+-/
+lemma pigeon_hole_principle (f : ℕ → color) (H : Inf):
+  infinite (reds f H) ∨ infinite (blues f H) :=
+begin
```

Comment:
I think that with the mathlib definition of `infinite` this proof can be 3 lines. `infininite := not finite` and union of finites is finite. Done...

Reply:
yep, that definition is great for the PHP. the other definition (the one I used) was convenient for actually selecting sufficiently large elements from a set. I guess the best thing to do would be to prove the equivalence of the definitions. And maybe that proof should live in the same file where mathlib's `infinite` is defined.